Expanded table render

kgaikwad opened the pull request:
Attaching below screenshot for reference:
![expanded_table](https://user-images.githubusercontent.com/6470528/101153432-73032d80-364a-11eb-9605-4140d90eafc9.png)


EDIT -

There are couple of things need to be fixed.
- [ ] Fix font Style,
- [x] color codes
- [x] Change `recommendations` column string values to link
- [x] Remove `%` with `out of 5` score
- [x] Resolve memory leak problem

terezanovotna commented:
Nice job Kavita!

Few other notes:
for the grading bar - the lighter color should be substituted with grey color as in [mocks](https://marvelapp.com/prototype/9408d9j/screen/74816602) - for the yellow example: darker yellow stays, the lighter yellow becomes grey

kgaikwad commented:
Hello @terezanovotna,
Added modifications to existing UI. Attaching the screenshot for reference.

![Screenshot from 2020-12-14 20-01-43](https://user-images.githubusercontent.com/6470528/102115783-08c86500-3e62-11eb-88c0-c83a227683f3.png)

Questions:
- Is font family same as the mock-up? if not could you share font details and sizes for table rows as well as for expanded view.
- could you confirm color codes for score bars and for links?
- I have removed recommendation button from expanded view as per mock-up. Is it expected, right?

terezanovotna commented:
Nice job Kavita! The font is same as in the mocks - Red Hat text. 
And yes, thanks for removing the button from the expanded area.

karelhala reviewed (commented):
@terezanovotna asked me to investigate the colors used in the progress, upon investigation I wouldn't use direct values, but rahter use progress variants. However if you want to fine tune the colors you should use css vars instead of direct values. If PF or anybody else decides to use different value value it will be automatically changed. Added few comments about class components as well, other than that looking good!

kgaikwad commented on src/Components/RosTable/RosTable.scss:
```diff
@@ -0,0 +1,78 @@
+  }
+}
+
+.green-100 {
+  .pf-c-progress__indicator {
+    background: #BDE2B9;
+  }
+}
+
+.green-400 {
+  .pf-c-progress__indicator {
+    background: #4CB140;
+  }
+}
+
+.gold-400 {
+  .pf-c-progress__indicator {
+    background: #F4C145;
+  }
+}
+
+.orange-300 {
+  .pf-c-progress__indicator {
+    background: #EC7A08;
+  }
+}
+
+.danger-200 {
+  .pf-c-progress__indicator {
+    background: #A30000;
+  }
+}
```
I tried to import below line and Progress component -

```
import { Progress, ProgressVariant } from '@patternfly/react-core';
<Progress value={100} title="Title" variant={ProgressVariant.success} />
```

It was giving me error - `unable to import ProgressVariant` so I just used Progress component.
Any suggestion on how to pass different color other than `'danger' | 'success' | 'warning'` using variant?

karelhala replied:
Mhh, I guess you are out of luck and have to either use classnames or ariables from `@patternfly/react-tokens`. Which you'll use is up to you, the tokens one would require setting style directly, which is not clean from CSS point, but all the logic would be in this file. With the classname you can use css vars for these as well, it will be cleaner from CSS side, but the logic would be spread in this file and in the stylesheet.

kgaikwad commented on src/Components/RosTable/ProgressScoreBar.js:
```diff
@@ -0,0 +1,36 @@
+import { Progress } from '@patternfly/react-core';
+import propTypes from 'prop-types';
+
+export const ProgressScoreBar = ({ valueScore, measureLocation }) =>  {
+    const calNumScore = Math.round(valueScore / 20);
+    const textLabel = `${ calNumScore }/5`;
+    const colorclass = (val) => {
+        switch (true) {
+            case (val === 5):
+                return 'green-400';
+            case (val === 4):
+                return 'green-100';
+            case (val === 3):
+                return 'gold-400';
+            case (val === 2):
+                return 'orange-300';
+            case (val <= 1):
+                return 'danger-200';
+        }
+    };
+
+    return (
+        <React.Fragment>
+            <Progress value={ calNumScore } min={0} max={5} label={ textLabel }
+                valueText={ textLabel }
+                className={ `progress-score-bar ${colorclass(calNumScore)}` }
+                measureLocation={ measureLocation } />
+        </React.Fragment>
+    );
+};
```
@karelhala, 
Here you suggested to make use of `variant` but if you see attached screenshot from [comment-744601282](https://github.com/RedHatInsights/ros-frontend/pull/4#issuecomment-744601282).

I need to use different shades of these colours (e.g. dark danger or light green) than existing variant. Is there any way I can pass colour intensity number (100/200/300/400) along with these variants?

kgaikwad replied:
Secondly, from server I am getting percentage value in value score.
After division by 20, I am considering round figure while showing a progress as well as text (i.e x out of 5). 
Example - result = 4.5 after division it will show 5/5 and 100% progress.

If I change code as per suggestion then there will be difference in progress (like for 95%, 98% and so on till 100%)  but text will show `5/5. It will be little confusing. Please let me know your thoughts on this.

kgaikwad commented:
[](url)
![Screenshot from 2020-12-28 17-50-37](https://user-images.githubusercontent.com/6470528/103214007-5c907f00-4935-11eb-87cd-5daa5b6e25cd.png)

@terezanovotna @karelhala this is updated screenshot after above changes. Please let me know if you have any additional comments/suggestions.

terezanovotna commented:
thanks @kgaikwad for the changes! from what I am seeing it looks really good!

Small nit pick - we are missing pagination

karelhala commented on src/Components/RosTable/ExpandedRow.js:
```diff
@@ -0,0 +1,41 @@
+    ];
+
+    return (
+        <React.Fragment>
```
Just a nitpick, no need to wrap in `React.Fragment` if you have just one child. But you can keep as is.

karelhala commented on src/Components/RosTable/ProgressScoreBar.js:
```diff
@@ -0,0 +1,36 @@
+export const ProgressScoreBar = ({ valueScore, measureLocation }) =>  {
+    const calNumScore = Math.round(valueScore / 20);
+    const textLabel = `${ calNumScore }/5`;
+    const colorclass = (val) => {
+        switch (true) {
+            case (val === 5):
+                return 'green-400';
+            case (val === 4):
+                return 'green-100';
+            case (val === 3):
+                return 'gold-400';
+            case (val === 2):
+                return 'orange-300';
+            case (val <= 1):
+                return 'danger-200';
+        }
+    };
```
I personaly prefer object mapper over switch, but that's just down to personal preferences.
```JS
const colorClass = (val) => ({
  5: 'green-400',
  4: 'green-100',
  3: 'gold-400',
  2: 'orange-300'
})?.[val] || 'danger-200'
```

kgaikwad replied:
Done! Modified code using object.

```
const colorclass = (val) =>  ({
        5: 'green-400',
        4: 'green-100',
        3: 'gold-400',
        2: 'orange-300'
    }[val] || 'danger-200');
```
Please let me know if any further changes required.

karelhala replied:
Looking good!

karelhala commented on src/Components/RosTable/RosTable.js:
```diff
@@ -0,0 +1,130 @@
+    <a href={ linkPath } className={ `pf-link ${classAsPerType} link-${id}` }>{textValue}</a>
+);
+
+const renderSystemLink = (id, textValue) => {
```
Instead of this function I'd use directly the `actionLink` function.

kgaikwad replied:
makes sense :+1:

karelhala commented on src/Components/RosTable/RosTable.js:
```diff
@@ -0,0 +1,130 @@
+                ];
+            });
+        } else {
+            let emptyRow = <EmptyTable>
```
Unless you want to change the empty row no need to assign to a value.

karelhala commented on src/Components/RosTable/RosTable.js:
```diff
@@ -0,0 +1,130 @@
+            });
+        } else {
+            let emptyRow = <EmptyTable>
+                <EmptyStateDisplay title={ 'No matching records found' } />
```
Direct string, no need to wrap in `{}`
```suggestion
                <EmptyStateDisplay title="No matching records found" />
```

karelhala commented:
Looking good! Just a few nitpicks and improvements @kgaikwad, nothing fundamentally wrong.

kgaikwad commented:
> Small nit pick - we are missing pagination

Yeah,  I have created a separate ticket to track pagination changes - [RHIROS-19](https://issues.redhat.com/browse/RHIROS-19).
Currently I am working on this ticket.

kgaikwad commented:
Hello @karelhala, 
Thank you for providing your inputs.Very helpful! :+1:  :ok_hand:
Updated changes as per suggestions. 
Please let me know if any further changes required.

kgaikwad commented:
Hello @terezanovotna, @karelhala,
do you have any additional comments?

@apuntamb @patilsuraj767 - please feel free to go through changes and let me know if any suggestions/comments.

karelhala commented:
Looking good! Thank you for taking my suggestions into consideration.

apuntamb reviewed (approved):
tested locally, works fine! :+1:

apuntamb commented:
Merged!
Thanks @kgaikwad for this great addition to UI. Thanks @terezanovotna for your guidance on how the UI should adhere to the designed mocks and big thanks to @karelhala for your thorough review and suggestions. :+1:

terezanovotna commented:
@katierik
